File: src/main/java/com/github/kpavlov/jreactive8583/netty/pipeline/EchoMessageListener.java
package com.github.kpavlov.jreactive8583.netty.pipeline;

import com.github.kpavlov.jreactive8583.IsoMessageListener;
import com.github.kpavlov.jreactive8583.iso.MessageClass;
import com.github.kpavlov.jreactive8583.iso.MessageFactory;
import com.solab.iso8583.IsoMessage;
import io.netty.channel.ChannelHandlerContext;

/**
 *
 */
public class EchoMessageListener<T extends IsoMessage> implements IsoMessageListener<T> {

    private final MessageFactory<T> isoMessageFactory;

    public EchoMessageListener(final MessageFactory<T> isoMessageFactory) {
        this.isoMessageFactory = isoMessageFactory;
    }

    @Override
    public boolean applies(final IsoMessage isoMessage) {
        return isoMessage != null && (isoMessage.getType() & MessageClass.NETWORK_MANAGEMENT.value()) != 0;
    }

    /**
     * Sends EchoResponse message. Always returns <code>false</code>.
     *
     * @param isoMessage a message to handle
     * @return <code>false</code> - message should not be handled by any other handler.
     */
    @Override
    public boolean onMessage(final ChannelHandlerContext ctx, final T isoMessage) {
        final IsoMessage echoResponse = isoMessageFactory.createResponse(isoMessage);
        ctx.writeAndFlush(echoResponse);
        return false;
    }
}

File: src/main/kotlin/com/github/kpavlov/jreactive8583/netty/pipeline/EchoMessageListener.kt
package com.github.kpavlov.jreactive8583.netty.pipeline

import com.github.kpavlov.jreactive8583.IsoMessageListener
import com.github.kpavlov.jreactive8583.iso.MessageClass
import com.github.kpavlov.jreactive8583.iso.MessageFactory
import com.solab.iso8583.IsoMessage
import io.netty.channel.ChannelHandlerContext

internal class EchoMessageListener<T : IsoMessage>(
    private val isoMessageFactory: MessageFactory<T>
) : IsoMessageListener<T> {

    override fun applies(isoMessage: T): Boolean {
        return isoMessage.type and MessageClass.NETWORK_MANAGEMENT.value != 0
    }

    /**
     * Sends EchoResponse message. Always returns `false`.
     *
     * @param isoMessage a message to handle
     * @return `false` - message should not be handled by any other handler.
     */
    override fun onMessage(ctx: ChannelHandlerContext, isoMessage: T): Boolean {
        val echoResponse: IsoMessage = isoMessageFactory.createResponse(isoMessage)
        ctx.writeAndFlush(echoResponse)
        return false
    }
}
